src/object/object_primitive.c

```diff
@@ -11339,21 +11487,87 @@ mr_data_cmpdisk_string (void *mem1, void *mem2, TP_DOMAIN * domain, int do_coerc
   assert (str_length1 == 0xFF || str_length2 == 0xFF);
 
   or_init (&buf1, str1, 0);
-  str_length1 = or_get_varchar_length (&buf1, &rc);
+  rc = or_get_varchar_compression_lengths (&buf1, &str1_compressed_length, &str1_decompressed_length);
+  if (rc != NO_ERROR)
+    {
+      goto cleanup;
+    }
+
+  string1 = db_private_alloc (NULL, str1_decompressed_length + 1);
+  if (string1 == NULL)
+    {
+      /* Error report */
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, str1_decompressed_length);
+      return rc;
+    }
+
+  rc = do_get_compressed_data_from_buffer (&buf1, string1, str1_compressed_length, str1_decompressed_length);
+  if (rc != NO_ERROR)
+    {
+      goto cleanup;
+    }
+  str_length1 = str1_decompressed_length;
+
   if (rc == NO_ERROR)
     {
```

Comment:
@lucianlazar2211 

Please strip `if` off. 

You may also optimize allocation and copying string to check its length.
```C
  if (str1_length == 0xFF)
    {
      ...
    }
  else
   {
     ...
     string1 = buf1.ptr;
     alloced_string1 = false;
    }
```
I don't think the optimization is mandatory.